OpenRA.Mods.Cnc/Traits/World/PavementRenderer.cs

```diff
@@ -0,0 +1,206 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Terrain;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[TraitLocation(SystemActors.World)]
+	[Desc("Attach this to the world actor. Renders the state of " + nameof(PavementLayer))]
+	public class PavementRendererInfo : TraitInfo
+	{
+		[Desc("Palette to render the layer sprites in.")]
+		public readonly string Palette = TileSet.TerrainPaletteInternalName;
+
+		public override object Create(ActorInitializer init) { return new BuildableAutoTerrainRenderer(init.Self, this); }
+	}
+
+	public class BuildableAutoTerrainRenderer : IRenderOverlay, IWorldLoaded, ITickRender, IRadarTerrainLayer, INotifyActorDisposing
+	{
+		readonly World world;
+		readonly PavementRendererInfo info;
+		readonly PavementLayer buildableAutoTerrainLayer;
+		readonly Dictionary<CPos, Sprite> dirty;
+		readonly ITiledTerrainRenderer terrainRenderer;
+		readonly CellLayer<(Color, Color)> radarColor;
+
+		TerrainSpriteLayer terrainSpriteLayer;
+		ITemplatedTerrainInfo terrainInfo;
+		PaletteReference paletteReference;
+
+		public BuildableAutoTerrainRenderer(Actor self, PavementRendererInfo info)
+		{
+			this.info = info;
+			world = self.World;
+			dirty = new Dictionary<CPos, Sprite>();
+			terrainRenderer = self.Trait<ITiledTerrainRenderer>();
+			buildableAutoTerrainLayer = self.Trait<PavementLayer>();
+			radarColor = new CellLayer<(Color, Color)>(world.Map);
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		{
+			paletteReference = wr.Palette(info.Palette);
+
+			terrainSpriteLayer = new TerrainSpriteLayer(w, wr, terrainRenderer.MissingTile, BlendMode.Alpha, wr.World.Type != WorldType.Editor);
+
+			if (!(w.Map.Rules.TerrainInfo is ITemplatedTerrainInfo terrainInfo))
```

Comment:
```suggestion
			if (!(w.Map.Rules.TerrainInfo is ITemplatedTerrainInfo templatedTerrainInfo))
```
to avoid "Local variable hides field" IDE warning, please.
Then you can also drop the `this.` two lines below.